src/util/psbt/mod.rs

```diff
@@ -88,6 +91,82 @@ impl PartiallySignedTransaction {
 
         Ok(())
     }
+
+    /// Decide whether the current Input spends segwit outputs or not and
+    /// calculate the corresponding Sighash.
+    /// #Panics:
+    /// Panics if the index >= number of inputs in psbt
+    pub fn signature_hash(&self, idx: usize) -> Result<SigHash, self::Error> {
+        let inp = &self.inputs[idx];
+        let sighash_type = inp.sighash_type.unwrap_or_else(|| SigHashType::All);
+        // Compute Script code for the Script. When this script is used as scriptPubkey, 
+        // the Script Code determines what goes into sighash.
+        // 1) For non-p2sh non-segwit outputs, script code is the public key.
+        let sighash = if let Some(ref non_witness_utxo) = inp.non_witness_utxo {
+            let spent_outpoint = self.global.unsigned_tx.input[idx].previous_output;
+            if spent_outpoint.txid != non_witness_utxo.txid() {
+                return Err(Error::InvalidNonWitnessUtxo {
+                    prevout_txid: spent_outpoint.txid,
+                    non_witness_utxo_txid: non_witness_utxo.txid()
+                });
+            }
+            let script_pubkey =  &non_witness_utxo.output[spent_outpoint.vout as usize].script_pubkey;
+            if let Some(ref redeem_script) = inp.redeem_script {
+                if redeem_script.to_p2sh() != *script_pubkey {
+                    return Err(Error::InvalidWitnessScript{
+                        expected: script_pubkey.clone().into_bytes(),
+                        actual: redeem_script.to_p2sh().into_bytes(),
+                    });
+                }
+                self.global.unsigned_tx.signature_hash(
+                    idx, &redeem_script, sighash_type.as_u32()
+                )
+            } else {
+                self.global.unsigned_tx.signature_hash(
+                    idx, &script_pubkey, sighash_type.as_u32()
+                )
+            }
+        } else if let Some(ref witness_utxo) = inp.witness_utxo {
+            let script_pubkey= if let Some(ref redeem_script) = inp.redeem_script {
+                if redeem_script.to_p2sh() != witness_utxo.script_pubkey {
+                    return Err(Error::InvalidWitnessScript{
+                        expected: witness_utxo.script_pubkey.clone().into_bytes(),
```

Comment:
Same down here, use `.to_bytes`

Reply:
Yeah, clone().into_bytes() looks silly. Idk Why I did this